x-pack/plugins/security_solution/server/lib/telemetry/sender.ts

```diff
@@ -109,6 +152,17 @@ export class TelemetryEventsSender {
     });
   }
 
+  // this returns a cached version of the telemetry opt-in. The cache is updated
+  // in the periodically executed sendIfDue function. This function returns false
+  // until the first execution of that function.
+  public isTelemetryOptedIn(): boolean {
+    if (!this.isOptedIn) {
+      return false;
+    } else {
+      return this.isOptedIn === true;
+    }
```

Comment:
Thanks! Will update.

Reply:
I noticed a state var to act as a cache from one of the collectors to the other creates a gnarly race condition. 
For example,

**State**: Opted out
```
-> Task Manager starts run
-> Discards alerts and returns until a detection signal is received by another process
-> User enables telemetry
```
**State**: Opted In
```
-> detection signal comes through, updates the opt in
-> User disables telemetry
```
**State**: Opted Out
```
-> diagnostic alerts still send
```

I think we should create another task that is responsible for polling the opt in status from the saved object and updates that var atomically - say every 30s/1m. That means there is only 1 process updating that var, the Diag Task manager and `sendIfDue` function are reading and there is no side effect related to opt in / opt out from those flows. We could also test this in isolation. What do you think, @tsg?

Reply:
@pjhampton Following this and noticed this conversation was resolved without any additional comments on this race condition... just curious if there was a resolution?

Looks like `isOptedIn` is still defaulted to `true` as well... I'm probably missing some context, but saw that it was suggested to default to `false` above.

Reply:
That's great. Thanks @madirey!  I was hoping to get your thoughts on jest testing this properly as I also used your task tests from the endpoint artifacts code as inspiration.

RE the race condition, this was based on the additional flag that was introduced here: https://github.com/elastic/kibana/pull/84268 - the reasoning behind this was that we didn't want to keep making a request to a saved object every minute (setInterval), and once again every 5 (task) but a discussion with @tsg offline (hence the resolve) was that we would just keep it simple as there were multiple mutators of the flag. It wasn't possible to ever set the flag if defaulted to false based on checking if the queue was empty in the `sendIfDue` function. Feedback / TS wisdom is welcome here :-)

Reply:
@pjhampton Sounds good! Thanks for the context! Just the one other comment about the `5m` time interval and possibly missing late events, but otherwise LGTM. 👍